### DISCUSSION TITLE
How to collect the script error in order to use that message instead of only getting an image?

### DISCUSSION THREAD

**Discussion by Illegit** [freeplane/freeplane#1618] [Help] [answered]
Is it possible to collect in clipboard or in another methodology the script error messages, like the one here?
![image](https://github.com/freeplane/freeplane/assets/62674157/aee5b7b6-44b2-4b60-890e-545d6cd0d831)
The image message is ok but I do not know if it is possible to copy it in other apps.
Thanks

**Comment by euu2021** [comment, marked as answer]
As lilive mentioned at https://github.com/freeplane/freeplane/discussions/1042#discussioncomment-5242229, the complete error message appears at the log.

Now, to make FP paste errors to the clipboard, one solution that I know is to use try/catch. Put your script inside a try/catch block:

```
try { 
  codeThatGivesError
} catch(Exception ex) {
  textUtils.copyToClipboard(ex.toString())
} 

```

**Reply by Illegit** [reply]
Thank you euu2021, I go with it.

**Reply by Illegit** [reply]
I refer to the #1042 from log file.: this i fine and enough for my use.
Just note that the script doesn't copy in my clipboard... please send any suggestion.

**Reply by euu2021** [reply]
> Just note that the script doesn't copy in my clipboard... please send any suggestion.

For example, if I execute this script:

```
try { 
  node.style.edge.colorCode = "#ffff00ff"
  lineOfCodeWithProblem
  node.style.edge.dash = "CLOSE_DOTS"
} catch(Exception ex) {
  textUtils.copyToClipboard(ex.toString())
} 
```
This happens:
- This line is applied to the map: `node.style.edge.colorCode = "#ffff00ff"`; 
- then this line gets an error: `lineOfCodeWithProblem`;
-  making the code jump to this line: `textUtils.copyToClipboard(ex.toString())`, which puts this in my clipboard: `groovy.lang.MissingPropertyException: No such property: lineOfCodeWithProblem for class: Script1`

**Reply by Illegit** [reply]
Hi euu2021, your script in your case is working even on my pc, but when I try the following one my clipboard remains empty: 


`try {

@Grab(group='org.jfree', module='jfreechart', version='1.5.4')
import groovy.swing.SwingBuilder
import org.jfree.chart.ChartFactory
import org.jfree.chart.ChartFrame
import org.jfree.chart.plot.PlotOrientation
import org.jfree.data.xy.XYSeriesCollection
import org.jfree.chart.labels.StandardXYItemLabelGenerator
import org.jfree.chart.renderer.xy.XYLineAndShapeRenderer
import org.jfree.chart.axis.NumberAxis

def selectedNode = getNode()
def title = selectedNode.text
def Xtext = selectedNode['Xtext']?.text ?: ""
def Ytext = selectedNode['Ytext']?.text ?: ""
def data = getData(selectedNode)
showChart(data, title, Xtext, Ytext, selectedNode)

// -------- methods -------------------

def createChart(data, title, Xtext, Ytext, selectedNode) {
    def dataset = new XYSeriesCollection()
    data.each { serieName, serieData ->
        def series = new org.jfree.data.xy.XYSeries(serieName)
        serieData.each { Xvalue, Yvalue ->
            series.add(Xvalue, Yvalue)
        }
        dataset.addSeries(series)
    }

    def chart = ChartFactory.createXYLineChart(
            title,
            Xtext,
            Ytext,
            dataset,
            PlotOrientation.VERTICAL,
            true,
            true,
            false
    )

    def renderer = new XYLineAndShapeRenderer(true, false)
    renderer.setDefaultItemLabelGenerator(new StandardXYItemLabelGenerator())
    renderer.setDefaultItemLabelsVisible(true)
    data.eachWithIndex { serieName, serieData, index ->
        def colorNode = selectedNode.children.find { it.text == serieName }['color']
        if (colorNode) {
            def color = java.awt.Color.decode(colorNode.to.text)
            renderer.setSeriesPaint(index, color)
        }
    }
    chart.getXYPlot().setRenderer(renderer)

    def xAxis = chart.getXYPlot().getDomainAxis()
    xAxis.setAutoRange(true)
    xAxis.setLabel(Xtext)

    def yAxis = chart.getXYPlot().getRangeAxis()
    yAxis.setAutoRange(true)
    yAxis.setLabel(Ytext)

    chart
}

def showChart(data, title, Xtext, Ytext, selectedNode) {
    def chart = createChart(data, title, Xtext, Ytext, selectedNode)
    def frame = new ChartFrame(title, chart)
    frame.pack()
    frame.setVisible(true)
}

def getData(node){
    def attrNameX = 'Xdata'
    def attrNameY = 'Ydata'
    def data = [:]
    node.children.each{serieNode ->
        def serieName = serieNode.text
        def serieData = [:]
        serieNode.children.sort { a, b -> a.name() <=> b.name() }.each{dataNode ->
            def Xvalue = dataNode[attrNameX]?.to?.num0
            def Yvalue = dataNode[attrNameY]?.to?.num0
            if (Xvalue != null && Yvalue != null) {
                serieData[Xvalue] = Yvalue
            }
        }
        data[serieName] = serieData
    }
    return data
}

def getNode(){
    def selectedNode = node
    selectedNode.children.first()
}

} catch(Exception ex) {
  textUtils.copyToClipboard(ex.toString())
} 
`

Please: note that I do not ask you to correct this wrong code, but only if you have an idea why my clipboard is still empty...thanks

**Reply by Illegit** [reply]
PS: I do not know why the full code doesn't stay in the code format...

**Reply by euu2021** [reply]
I have no idea. I see that one of the problems is that the try/catch can't have imports inside it. Explanation: https://stackoverflow.com/questions/68941420/is-there-a-way-to-place-a-try-catch-block-around-an-import-in-groovy

**Comment by i-plasm** [comment]
Hi @Illegit  . I wrote a script that is able to detect whenever a Freeplane error dialog appears, and allows the user to copy the error to clipboard. Right now it is configured to detect the error dialogs whose title is "Freeplane". 

![errordialogdetection](https://github.com/freeplane/freeplane/assets/144830522/67dbca88-9aa8-49e1-94cb-bd5bdd222a90)

The script is found [here](https://raw.githubusercontent.com/i-plasm/freeplane-scripts/main/src/scripts/freeplaneErrorDialog2Clipboard.groovy). It works either as a regular or init script. You can find more instructions in the script header. It works for me, let me know if you try it.

**Comment by Illegit** [comment]
Hi I-plasm: I used your scripts and created a menu (Edo style): 
![image](https://github.com/freeplane/freeplane/assets/62674157/81d63652-944f-40f6-8f29-912f0ca57a82)
Thank you, very appreciated.

**Reply by i-plasm** [reply]
Glad it helped!
